server/services/participants.js

```diff
@@ -211,6 +149,146 @@ const withdrawParticipant = async (participantInfo) => {
+    });
+
+    // Add an ephemeral status to warn the employer
+    await tx[collections.PARTICIPANTS_STATUS].save({
+      employer_id: chosenOne.employer_id,
+      participant_id: participantId,
+      status: 'pending_acknowledgement',
+      current: true,
+      data,
+    });
+    await withdrawParticipant(chosenOne);
+  });
+};
+
+const setParticipantStatus = async (
+  employerId,
+  participantId,
+  status,
+  data, // JSONB on the status row
+  isHa = false
+) =>
+  dbClient.db.withTransaction(async (tx) => {
+    if (status === 'pending_acknowledgement') {
+      return { status: 'invalid_status' };
+    }
+    if (status !== 'rejected' && status !== 'archived') {
+      const items = await tx[collections.PARTICIPANTS_STATUS].find({
+        participant_id: participantId,
+        status: 'hired',
+        current: true,
+      });
+
+      if (items.length > 0) return { status: 'already_hired' };
+    }
+
+    const item = await tx[collections.PARTICIPANTS_STATUS].findOne({
+      participant_id: participantId,
+      employer_id: employerId,
+      current: true,
+    });
+    // Check the desired status against the current status:
+    // -- Rejecting a participant is allowed even if they've been hired elsewhere (handled above)
+    // -- Open is the starting point, there is no way to transition here from any other status
+    // -- If engaging (prospecting), participant must be coming from open, null, or rejected status
+    // -- If interviewing, participant must be coming from prospecting status
+    // -- If offer made, must be coming from interviewing
+    // -- If hiring, must be coming from offer made
+    // -- If restoring a user from being archived, any status should be valid
+    if (
+      (status === 'open' ||
+        (status === 'prospecting' &&
+          item !== null &&
+          item.status !== 'open' &&
+          item.status !== 'rejected') ||
+        (status === 'interviewing' && item?.status !== 'prospecting') ||
+        (status === 'offer_made' && item?.status !== 'interviewing') ||
+        (status === 'hired' && item?.status !== 'offer_made')) &&
+      item?.status !== 'archived'
+    )
+      return { status: 'invalid_status_transition' };
+    if (status === 'archived' && !item) {
+      return { status: 'Could not find participant' };
+    }
+    // Invalidate pervious status
+    await tx[collections.PARTICIPANTS_STATUS].update(
+      {
+        employer_id: employerId,
+        participant_id: participantId,
+        current: true,
+      },
+      { current: false }
+    );
+
+    // Save new status
+    await tx[collections.PARTICIPANTS_STATUS].save({
+      employer_id: employerId,
+      participant_id: participantId,
+      status,
+      current: true,
+      data,
+    });
+
+    const participant = await tx[collections.PARTICIPANTS].findDoc({
+      id: participantId,
+    });
+    // Now check if current status is archived then set interested flag
+    if (status === 'archived') {
```

Comment:
Now there is a dedicated method to archive participant. we should remove this status check. And put code under if to archive participant method. Any thought.

Reply:
I have mixed feelings about this. One of the benefits of using the changeStatus route for archiving is that the user can only archive their own participants.